### ISSUE TITLE
Class selectors in nested rules

### ISSUE THREAD

**Issue by jankrcal** [less/less.ruby#109]
I want to set rules for `A`, `A:hover`, `A.highlighted` and `A *.child-element`. It all relates to `A` so it makes sense to write it all inside a nested rule as follows

```
A {
    color: blue;
    :hover {
        color: black;
     }
    .highlighted {
        color: red;
    }
    *.child-element {
        color: gray;
    }
}
```

But this gets compiled into
    ...
    A .highlighted {color: red;}
    A *.child-element {color: gray;}

I propose that any nested selector without the name specified (e.g. `.class-name`, `:pseudo-class`, `[attribute="value"]`) relates to the parent element. At the moment, this feature works for pseudo classes but not for other selectors.

I know this proposal has one major flaw - it is not backward compatible. The backward compatible but IMHO uglier alternative is to introduce a special name selector that refers to the parent. E.g. `self` or `$`. So one would have to write `self.highlighted` or `$.highlighted`.

What you think?

**Comment by jtousek** [comment]
I must agree with that. Let's look to this code.:

<pre>
A {
    :hover { color: black; }
    .highlighted { color: red; }
}
</pre>

compiled:

<pre>
A:hover { color: black; }
A .highlighted { color: red; }
</pre>

In first case there is no space, but what if a want (theoreticly) this?

<pre>
A :hover { color: black; }
</pre>

Well there is no chance - but i can use *:hover. - OK but the difference with space between .class and :pseudo-class is not logical.
And I never can get something like this using mixin for A

<pre>
A:hover { color: black; }
A.highlighted { color: red; } //without space
</pre>

I think that because of incompatibility this should be changed in new major version - LESS 2.0. But I vote for this solution.:

<pre>
A {
    :hover { color: black; }
    .highlighted { color: red; }
}
</pre>

compiled to 

<pre>
A :hover { color: black; }
A .highlighted { color: red; }
</pre>

and

<pre>
A {
    $:hover { color: black; }
    $.highlighted { color: red; }
}
</pre>

compiled to 

<pre>
A:hover { color: black; }
A.highlighted { color: red; }
</pre>

The first case is used much more often - that's why I suggest this.

**Comment by cloudhead** [comment]
`.foo.bar` is the same thing as `.bar.foo`, for that reason, it doesn't make as much sense to nest either. pseudo classes like `:hover` don't work that way.

**Comment by jankrcal** [comment]
I don't get your point. Isn't `.foo:hover` the same as `:hover.foo`? 

The reason for me is the clarity and elegance of the code. With my proposal, one can place all rules concerning **one** element into **one** block.

**Comment by jtousek** [comment]
clouhead: Yes of course, but what about this?

<pre>
#header{
//global settings for page header
$.page1{
//special settings for page 1
}
$.page2{
//special settings for page 2
}
}
</pre>

Now I have to write this:

<pre>
#header{
//global settings for page header
}
#header.page1{
//special settings for page 1
}
#header.page2{
//special settings for page 2
}
</pre>

I think you should agree that in this case the old way is not good enough.

**Comment by cloudhead** [comment]
Ok, you convinced me : )

I'm thinking of something like this for the syntax:
    .foo {
        `.bar {...} // .foo.bar
        .baz {...} // .foo .baz
    }

any objections? (it's a backtick) pseudo-classes won't change for now though, as that would break things, you can still have it both ways by doing `*:hover {...}`

**Comment by jtousek** [comment]
well I really don't like backtick (`), I'd prefere $ or \ or &...
about pseoudo-classes and pseudo-elements I agree

**Comment by jankrcal** [comment]
great :)

what about `<`? As a kind of opposite of the direct ancestor concatenator. But I don't mind if it is the backtick.

**Comment by jankrcal** [comment]
Example:

```
DIV {
    <.class1 {...}
    <.class2 {...}
}
```

**Comment by jtousek** [comment]
I'm still against backtick. :) But I'm ok with < idea (of course $ is prefered). Let's wait for cloudhead's opinion.

**Comment by cloudhead** [comment]
```
&.class
```

\+ same as Sass  
- quite noisy

```
<.class
```

\+ could make sense as an opposit to `>`  
- looks a little funny (nothing like it in any other language)

```
$.class
```

\+ `$` is unique enough  
- no obvious logic behind it, a little noisy

```
`.class
```

\+ very discreet  
- people don't like backticks

**Comment by jtousek** [comment]
I agree but people have to get used to it no matter which one it'll be. Another option might be %.  Well one has to be chosen and imho $ is still the best choice even if it was not originaly my but jankrcal's idea. Obvious logic isn't behind any of them (not only $ - maybe except <). Backtick in any code looks just like some wierd char which should be deleted. ;)

**Comment by thisisfranciswu** [comment]
Ampersand.  For serious.  Please.  With a cherry on top.  I'm confused enough as it is switching back and forth between SASS and LESS.

**Comment by jtousek** [comment]
considered that, I guess & is the best after all, it is good if it's similar to SASS (even if I have no experience with it)

**Comment by DouweM** [comment]
I vote for `&` or `$`. `&` would be nice because it's like in SASS, and `$` would rock, just because it's `$`.

**Comment by jeremyh** [comment]
This is great. The inconsistency between pseudo-elements and other simple-selectors (like attribute selectors) has always bugged me.

I like the backticks. They aren't as visually noisy as the other characters.

**Comment by jeremyh** [comment]
I've removed the paragraph. In hindsight it wasn't worth mentioning.

**Comment by tatey** [comment]
`<` feels the most intuitive for referencing the parent element. For people familiar with CSS and unfamiliar with LESS, they're going to think of `<` as the opposite of the child selector.

**Comment by jtousek** [comment]
Agreed. I think it is between `<` and `&` now. Cloudhead?

**Comment by DouweM** [comment]
I vote for `&`, as it functions more as a `this` or `self` selector than a parent selector... So `<` wouldn't actually be the opposite of the `>` selector.

**Comment by cloudhead** [comment]
`&` it is!

The thing to remember is `.a { &.b {...} }` is the same as `.b { &.a {...} }` so it's more of an 'append' than simply a reference to the parent selector. I liked the backticks, but they aren't that obvious, and should probably be kept for quoting things.

**Comment by DouweM** [comment]
Yeah, but `.a { ... &.b { ... } }` isn't the same as `.b { ... &.a { ... } }` ;-) But I get what you mean :-)

(Will  `.a { &:hover { ... } }` also be available next to the original `.a { :hover { ... } }`?)

**Comment by cloudhead** [comment]
Yea, `&:hover` should behave the same as `:hover`. If you want the equivalent of `a :hover`, you can use `*:hover`

**Comment by DouweM** [comment]
Ok :-) And is it already in 1.2.20?

**Comment by jtousek** [comment]
great! :)
btw. cloudhead, is there some kind of change log? there are lot of issues reported here but I don't know what is already implemented/fixed and what is still waiting...

**Comment by cloudhead** [comment]
@Douwe: I haven't implemented this yet, no : )

@jtousek: if the issue is still open here, then it's not fixed/implemented. The most accurate changelog would probably be the commit history, although that can be hard to follow — I don't really have the patience/time to maintain a separate one...

**Comment by jtousek** [comment]
cloudhead: Yeah, just as I thought... :-D

**Comment by wmdmark** [comment]
Any ETA on a fix for this?

**Comment by protyposis** [comment]
Just signed up to thank you for your fantastic work... and to note that this missing feature is what prevents me from using less :(

**Comment by DouweM** [comment]
Anything?

**Comment by jtousek** [comment]
Well what prevents me from using less is this thing along with "filter: ..." incompatibility... I'm looking forward for finishing these issues.

**Comment by jeremyh** [comment]
Use Sass if you need the features now: http://sass-lang.com/

**Comment by jtousek** [comment]
I just hate the syntax of Sass. I know LESS has some key features missing, but it's still better.

**Comment by jtousek** [comment]
cloudhead: one more thing to this one:

<pre>
input{
//global settings for inputs
&[type=text]{
//special settings for text inputs
}
&[type=submit]{
//special settings for submits
}
&[type=checkbox],
&[type=radio]{
//special settings for checkbox and radio inputs
}
}
</pre>

**Comment by nogweii** [comment]
Just throwing my 2 cents in, but why not have LESS be smart about the space? And, avoid the self-reference character. If you have `.class { .subclass { ... } }` then LESS should produce `.class .subclass { ... }` - but if it's an element reference (detected by the block not beginning with anything) it should produce `A.subclass { ... }`.

**Comment by jankrcal** [comment]
evaryont: it would break the backwards compatibility! 
Furthermore, I think that it is better to write one more character and have a simply defined behavior. There are additional corner-case situations where you have not defined the "smart" way.

**Comment by cloudhead** [comment]
I implemented this in less.js, which will be stable soon, it works great.

http://github.com/cloudhead/less.js
